src/dasi/shelf/ShelfCatalogue.cc

```diff
@@ -0,0 +1,74 @@
+
+#include "dasi/shelf/ShelfCatalogue.h"
+
+#include "dasi/api/Config.h"
+
+#include "dasi/core/FileReadHandle.h"
+#include "dasi/core/SplitReferenceKey.h"
+
+#include <cctype>
+#include <fstream>
+#include <iomanip>
+#include <sstream>
+
+namespace dasi::shelf {
+
+static std::string encode(const std::string_view& v) {
+    std::ostringstream r;
+    r.fill('0');
+    r << std::hex;
+
+    for (const auto c : v) {
+        auto cu = static_cast<unsigned char>(c);
+        if (std::isalnum(cu) || c == '-' || c == '.' || c == '_')
+            r << c;
+        else
+            r << ':' << std::setw(2) << int(cu);
+    }
+
+    return r.str();
+}
+
+ShelfCatalogue::ShelfCatalogue(const core::OrderedReferenceKey& dbkey, const api::Config& config) :
+    Catalogue(dbkey, config),
+    root_(config.getString("root")) {}
+
+
+void ShelfCatalogue::archive(const core::SplitReferenceKey& key, const void* data, size_t length) {
+    ASSERT(key[0] == dbkey());
+    auto path = buildPath(key);
+    std::filesystem::create_directories(path.parent_path());
+    std::ofstream fout(path, std::ios::out | std::ios::binary);
+    fout.write(reinterpret_cast<const char*>(data), length);
```

Comment:
Looking at this raises the question - what are the flush() semantics?

Reply:
We won't be able to do much more than the close() in the ofstream destructor, as it is now. I just went for the simplest implementation for now, to be able to test the interface